packages/jupyter-ai-magics/jupyter_ai_magics/magics.py

```diff
@@ -577,13 +597,29 @@ def run_ai_cell(self, args: CellArgs, prompt: str):
         ip = self.shell
         prompt = prompt.format_map(FormatDict(ip.user_ns))
 
+        context = self.transcript[-2 * self.max_history :] if self.max_history else []
         if provider.is_chat_provider:
-            result = provider.generate([[HumanMessage(content=prompt)]])
+            result = provider.generate([[*context, HumanMessage(content=prompt)]])
         else:
             # generate output from model via provider
-            result = provider.generate([prompt])
+            if self.transcript:
```

Comment:
The condition should be based on whether `context` exists, since `context` is used to build the input array, not `self.transcript`:

```suggestion
            if context:
```